File: NestCash/backend/app/models/csv_import_models.py
# app/models/csv_import_models.py
from pydantic import BaseModel, Field, validator
from typing import List, Dict, Optional, Any
from datetime import datetime
from enum import Enum

class CSVColumnType(str, Enum):
    """CSV oszlop típusok"""
    DATE = "date"
    AMOUNT = "amount"
    DESCRIPTION = "description"
    TYPE = "type"
    CURRENCY = "currency"
    CATEGORY = "category"
    IGNORE = "ignore"

class ColumnMapping(BaseModel):
    """Egy CSV oszlop leképezése az app mezőjére"""
    csv_column_name: str = Field(..., description="CSV oszlop neve")
    app_field: CSVColumnType = Field(..., description="App mező típusa")
    required: bool = Field(True, description="Kötelező mező-e")

class CSVPreviewRow(BaseModel):
    """CSV előnézet egy sora"""
    row_index: int
    data: Dict[str, Any]
    parsed_data: Optional[Dict[str, Any]] = None
    errors: List[str] = Field(default_factory=list)

class CSVPreviewResponse(BaseModel):
    """CSV előnézet válasz"""
    headers: List[str]
    sample_rows: List[CSVPreviewRow]
    total_rows: int
    detected_mappings: List[ColumnMapping] = Field(default_factory=list)

class ImportConfiguration(BaseModel):
    """Import konfigurációs beállítások"""
    main_account: str = Field(..., description="Főszámla típusa")
    sub_account_name: str = Field(..., description="Alszámla neve")
    default_category: Optional[str] = Field(None, description="Alapértelmezett kategória")
    column_mappings: List[ColumnMapping] = Field(..., description="Oszlop leképezések")
    skip_duplicates: bool = Field(True, description="Duplikátumok átugrása")
    date_format: str = Field("%Y-%m-%d %H:%M:%S", description="Dátum formátum")

class TransactionImportData(BaseModel):
    """Egy importálandó tranzakció adatai"""
    date: str
    amount: float
    description: str
    type: str  # income/expense
    category: Optional[str] = None
    currency: str = "HUF"
    is_duplicate: bool = False
    original_row: Dict[str, Any]

class ImportExecuteRequest(BaseModel):
    """Import végrehajtási kérés"""
    file_data: str = Field(..., description="Base64 encoded CSV data")
    configuration: ImportConfiguration

class ImportResult(BaseModel):
    """Import eredmény"""
    success_count: int = 0
    error_count: int = 0
    duplicate_count: int = 0
    errors: List[Dict[str, Any]] = Field(default_factory=list)
    imported_transaction_ids: List[str] = Field(default_factory=list)

class ImportStats(BaseModel):
    """Import statisztikák"""
    total_income: float = 0.0
    total_expenses: float = 0.0
    by_category: Dict[str, float] = Field(default_factory=dict)
    by_currency: Dict[str, float] = Field(default_factory=dict)

File: NestCash/frontend/lib/models/csv_import_models.dart
// lib/models/csv_import_models.dart

enum CSVColumnType {
  date('date'),
  amount('amount'),
  description('description'),
  type('type'),
  currency('currency'),
  category('category'),
  ignore('ignore');

  const CSVColumnType(this.value);
  final String value;

  static CSVColumnType fromString(String value) {
    return values.firstWhere((e) => e.value == value, orElse: () => ignore);
  }
}

class ColumnMapping {
  final String csvColumnName;
  final CSVColumnType appField;
  final bool required;

  ColumnMapping({
    required this.csvColumnName,
    required this.appField,
    required this.required,
  });

  factory ColumnMapping.fromJson(Map<String, dynamic> json) {
    return ColumnMapping(
      csvColumnName: json['csv_column_name'] as String,
      appField: CSVColumnType.fromString(json['app_field'] as String),
      required: json['required'] as bool,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'csv_column_name': csvColumnName,
      'app_field': appField.value,
      'required': required,
    };
  }

  ColumnMapping copyWith({
    String? csvColumnName,
    CSVColumnType? appField,
    bool? required,
  }) {
    return ColumnMapping(
      csvColumnName: csvColumnName ?? this.csvColumnName,
      appField: appField ?? this.appField,
      required: required ?? this.required,
    );
  }
}

class CSVPreviewRow {
  final int rowIndex;
  final Map<String, dynamic> data;
  final Map<String, dynamic>? parsedData;
  final List<String> errors;

  CSVPreviewRow({
    required this.rowIndex,
    required this.data,
    this.parsedData,
    required this.errors,
  });

  factory CSVPreviewRow.fromJson(Map<String, dynamic> json) {
    return CSVPreviewRow(
      rowIndex: json['row_index'] as int,
      data: Map<String, dynamic>.from(json['data'] as Map),
      parsedData: json['parsed_data'] != null 
          ? Map<String, dynamic>.from(json['parsed_data'] as Map)
          : null,
      errors: List<String>.from(json['errors'] as List),
    );
  }

  bool get hasErrors => errors.isNotEmpty;
}

class CSVPreviewResponse {
  final List<String> headers;
  final List<CSVPreviewRow> sampleRows;
  final int totalRows;
  final List<ColumnMapping> detectedMappings;

  CSVPreviewResponse({
    required this.headers,
    required this.sampleRows,
    required this.totalRows,
    required this.detectedMappings,
  });

  factory CSVPreviewResponse.fromJson(Map<String, dynamic> json) {
    return CSVPreviewResponse(
      headers: List<String>.from(json['headers'] as List),
      sampleRows: (json['sample_rows'] as List)
          .map((row) => CSVPreviewRow.fromJson(row as Map<String, dynamic>))
          .toList(),
      totalRows: json['total_rows'] as int,
      detectedMappings: (json['detected_mappings'] as List)
          .map((mapping) => ColumnMapping.fromJson(mapping as Map<String, dynamic>))
          .toList(),
    );
  }
}

class ImportConfiguration {
  final String mainAccount;
  final String subAccountName;
  final String? defaultCategory;
  final List<ColumnMapping> columnMappings;
  final bool skipDuplicates;
  final String dateFormat;

  ImportConfiguration({
    required this.mainAccount,
    required this.subAccountName,
    this.defaultCategory,
    required this.columnMappings,
    this.skipDuplicates = true,
    this.dateFormat = '%Y-%m-%d %H:%M:%S',
  });

  Map<String, dynamic> toJson() {
    return {
      'main_account': mainAccount,
      'sub_account_name': subAccountName,
      'default_category': defaultCategory,
      'column_mappings': columnMappings.map((m) => m.toJson()).toList(),
      'skip_duplicates': skipDuplicates,
      'date_format': dateFormat,
    };
  }
}

class ImportResult {
  final int successCount;
  final int errorCount;
  final int duplicateCount;
  final List<Map<String, dynamic>> errors;
  final List<String> importedTransactionIds;

  ImportResult({
    required this.successCount,
    required this.errorCount,
    required this.duplicateCount,
    required this.errors,
    required this.importedTransactionIds,
  });

  factory ImportResult.fromJson(Map<String, dynamic> json) {
    return ImportResult(
      successCount: json['success_count'] as int,
      errorCount: json['error_count'] as int,
      duplicateCount: json['duplicate_count'] as int,
      errors: List<Map<String, dynamic>>.from(json['errors'] as List),
      importedTransactionIds: List<String>.from(json['imported_transaction_ids'] as List),
    );
  }

  bool get hasErrors => errorCount > 0 || errors.isNotEmpty;
  int get totalProcessed => successCount + errorCount + duplicateCount;
}

class UserImportData {
  final List<String> categories;
  final Map<String, List<String>> subAccounts;
  final List<String> supportedCurrencies;

  UserImportData({
    required this.categories,
    required this.subAccounts,
    required this.supportedCurrencies,
  });

  factory UserImportData.fromJson(Map<String, dynamic> json) {
    final accounts = json['accounts'] as Map<String, dynamic>;
    final subAccountsMap = accounts['sub_accounts'] as Map<String, dynamic>;
    
    return UserImportData(
      categories: List<String>.from(json['categories'] as List),
      subAccounts: subAccountsMap.map((key, value) => 
          MapEntry(key, List<String>.from(value as List))),
      supportedCurrencies: List<String>.from(json['supported_currencies'] as List),
    );
  }

  List<String> get mainAccounts => subAccounts.keys.toList();
}